src/tracked_struct.rs

```diff
@@ -231,8 +291,35 @@ where
 }
 // ANCHOR_END: ValueStruct
 
-#[derive(Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Copy, Clone)]
-pub struct Disambiguator(pub u32);
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Copy, Clone)]
+pub struct Disambiguator(u32);
+
+#[derive(Default, Debug)]
+pub(crate) struct DisambiguatorMap {
+    // we use a non-hasher hashmap here as our key contains its own hash (in a sense)
+    // so we use the raw entry api instead to avoid the overhead of hashing unnecessarily
+    map: hashbrown::HashMap<IdentityHash, Disambiguator, ()>,
+}
+
+impl DisambiguatorMap {
+    pub(crate) fn disambiguate(&mut self, key: IdentityHash) -> Disambiguator {
+        use hashbrown::hash_map::RawEntryMut;
+
+        let eq_modulo_hash = |k: &IdentityHash| k.ingredient_index == key.ingredient_index;
```

Comment:
It seems that we also need to include `k.hash == key.hash`....

Reply:
To add a little context, hashbrown doesn't store a full hash of each key, and only compares partially. The equality function cannot rely on the hash being equal, it must uniquely identify an entry regardless (e.g. hashbrown could legally do a linear search for equality).

Reply:
Ah I see, well that would do it ... apologies